contract_forecast/models/contract_line.py

```diff
@@ -69,38 +69,41 @@ def _generate_forecast_periods(self):
         for rec in self:
             rec.forecast_period_ids.unlink()
             if rec.recurring_next_date:
-                last_date_invoiced = (
+                period_date_end = (
                     rec.last_date_invoiced
                     if rec.last_date_invoiced
                     else rec.date_start - relativedelta(days=1)
                 )
-                period_date_end = last_date_invoiced
-                recurring_next_date = rec.recurring_next_date
-                while rec._get_generate_forecast_periods_criteria(
+                while (
                     period_date_end
-                ):
-                    period_dates = rec._get_period_to_invoice(
-                        last_date_invoiced,
-                        recurring_next_date,
-                        stop_at_date_end=not rec.is_auto_renew,
+                    and rec._get_generate_forecast_periods_criteria(
+                        period_date_end
                     )
-                    period_date_start, period_date_end, recurring_next_date = (
-                        period_dates
+                ):
```

Comment:
I think this can be simplified by first initializing `recurring_next_date = rec.recurring_next_date; period_date_end = rec.next_period_date_end; period_date_start = rec.next_period_date_start` and then removing `if rec.recurring_next_date` and moving that test in the `while` condition. This is not blocking for this PR and can be done in a follow-up.